Sources/Datadog/Core/System/NetworkConnectionInfoProvider.swift

```diff
@@ -35,54 +35,152 @@ internal struct NetworkConnectionInfo {
+/// The `ThreadSafeNWPathMonitor` listens to path updates and synchonizes the values on `.current` property.
+/// This adds the necessary thread-safety and keeps the convenience of pulling.
+@available(iOS 12, *)
+internal class NWPathNetworkConnectionInfoProvider: NetworkConnectionInfoProviderType {
+    /// Queue synchronizing the reads and updates to `NWPath`.
+    private let queue = DispatchQueue(
+        label: "com.datadoghq.thread-safe-nw-path-monitor",
+        target: .global(qos: .utility)
+    )
+    private let monitor: NWPathMonitor
+
+    init(monitor: NWPathMonitor = NWPathMonitor()) {
+        self.monitor = monitor
+        monitor.pathUpdateHandler = { [weak self] path in
+            self?.unsafeCurrentNetworkConnectionInfo = NetworkConnectionInfo(
+                reachability: NetworkConnectionInfo.Reachability(from: path.status),
+                availableInterfaces: Array(fromInterfaceTypes: path.availableInterfaces.map { $0.type }),
+                supportsIPv4: path.supportsIPv4,
+                supportsIPv6: path.supportsIPv6,
+                isExpensive: path.isExpensive,
+                isConstrained: {
+                    if #available(iOS 13.0, *) {
+                        return path.isConstrained
+                    } else {
+                        return nil
+                    }
+                }()
+            )
+        }
+        monitor.start(queue: queue)
+    }
+
+    deinit {
+        monitor.cancel()
+    }
+
+    /// Unsynchronized `NetworkConnectionInfo`. Use `self.current` setter & getter.
+    private var unsafeCurrentNetworkConnectionInfo: NetworkConnectionInfo?
+    var current: NetworkConnectionInfo? {
+        get { queue.sync { unsafeCurrentNetworkConnectionInfo } }
+        set { queue.async { self.unsafeCurrentNetworkConnectionInfo = newValue } }
+    }
+}
+
+// MARK: - iOS 11
+
+import SystemConfiguration
+
+internal class iOS11NetworkConnectionInfoProvider: NetworkConnectionInfoProviderType {
+    private let reachability: SCNetworkReachability = {
+        var zero = sockaddr()
+        zero.sa_len = UInt8(MemoryLayout<sockaddr>.size)
+        zero.sa_family = sa_family_t(AF_INET)
+
+        return SCNetworkReachabilityCreateWithAddress(nil, &zero)! // swiftlint:disable:this force_unwrapping
+    }()
+
+    var current: NetworkConnectionInfo? {
+        var retrieval = SCNetworkReachabilityFlags()
+        let flags = (SCNetworkReachabilityGetFlags(reachability, &retrieval)) ? retrieval : nil
+        return NetworkConnectionInfo(
+            reachability: NetworkConnectionInfo.Reachability(from: flags),
+            availableInterfaces: Array(fromReachabilityFlags: flags),
+            supportsIPv4: nil,
+            supportsIPv6: nil,
+            isExpensive: nil,
+            isConstrained: nil
         )
-        self.queue = queue
-        self.fetchBlock = {
-            return provider.current
```

Comment:
As far as I understand, the problem is that `NWPathMonitor` writes into `currentPath` in its `queue` and we read it from other queues which leads to the crash because of disintegrated `currentPath` object.

If that's correct, adding `queue.sync { return provider.current }` to this line should solve it. 

At first I thought using `pathHandler` could be a good idea too but now it seems like it only complicates the problem, it gives a mutable `unsafeCurrentNetworkConnectionInfo` that we don't seem to need

In fact, maybe we don't even need to capture reader `queue` separately, we can do
```swift
self.fetchBlock = { [capturedMonitor = provider] in
  capturedMonitor.queue.sync { return capturedMonitor.currentPath }
}
```
What do you think?

Reply:
> If that's correct, adding `queue.sync { return provider.current }` to this line should solve it.

Synchronizing the `provider.current` on the same queue as we pass to `NWPathMonitor` is exactly what I did in the first iteration of this PR ([first 3 commits](https://github.com/DataDog/dd-sdk-ios/pull/109/files/87ee84bf2441ed3313b5d1796aedcb83af090a6a)). But later, I thought that this is somehow depending on the implementation detail of the `NWPathMonitor`. Yes, the queue used for `pathHandler` seems to be the same which writes the `.currentPath` in [this part of `NWPathMonitor` implementation](https://github.com/apple/swift/blob/bbe2f6e735e98c7b1d21aa91b804705e750a532f/stdlib/public/Darwin/Network/NWPath.swift#L284-L289
):

```swift
nw_path_monitor_set_update_handler(self.nw) { (newPath) in
   self.currentPath = NWPath(newPath)
   if let handler = self._pathUpdateHandler {
      handler(self.currentPath)
   }
}
```

but what I see from the public API docs is:
```swift
/// Start the path monitor and set a queue on which path updates
/// will be delivered.
public func start(queue: DispatchQueue)

/// Get queue used for delivering the pathUpdateHandler block.
public var queue: DispatchQueue?
```
vs:
```swift
/// Access the current network path tracked by the monitor
public var currentPath = NWPath(nil)
```

Which means:
* clear threading information for `pathUpdateHandler` in public API doc;

vs:

* an implicit promise on `.currentPath` threading model based on the current implementation.

Reply:
1. `NWPathMonitor.currentPath` is a readonly property (it's a `var` but `NWPath` doesn't have `public init`)
2. As it sets `currentPath` internally, it should've made sure its getter/setter were using the same thread
3. But it doesn't do so and we need to deal with it

To my understanding, there is clearly a problem with `NWPathMonitor` implementation. It wasn't even needed to state thread safety of `currentPath` explicitly, it was supposed to be by default.

Anyway, that `pathUpdateHandler` work does add a bit of complexity but it's still at an acceptable level ✅ 
Let's keep it that way and see how it goes 🤞